File: problems/coding/019-course-schedule/solution.py
def canFinish(n, prereqs):
    from collections import defaultdict, deque
    g = defaultdict(list); indeg = [0]*n
    for a, b in prereqs:
        g[b].append(a); indeg[a] += 1
    q = deque([i for i in range(n) if indeg[i] == 0]); seen = 0
    while q:
        v = q.popleft(); seen += 1
        for w in g[v]:
            indeg[w] -= 1
            if indeg[w] == 0: q.append(w)
    return seen == n

File: problems/coding/019-course-schedule/solution.js
function canFinish(n, prereqs) {
  const g = Array.from({length: n}, () => []); const indeg = Array(n).fill(0);
  for (const [a, b] of prereqs) { g[b].push(a); indeg[a]++; }
  const q = []; for (let i = 0; i < n; i++) if (indeg[i] === 0) q.push(i);
  let seen = 0;
  while (q.length) { const v = q.shift(); seen++; for (const w of g[v]) if (--indeg[w] === 0) q.push(w); }
  return seen === n;
}
